### PR TITLE
[ENH] New transformation based pipeline classifiers

### LINKED ISSUE #1063: [ENH] time series classification: shorthands for common sktime pipelines including TsfreshClassifier
One of our most requested/used time series classifiers is the pipeline of `tsfresh` feature extraction and then an `sklearn` classifier. 
- [x] There should be a shorthand `TsfreshClassifier` for that.

We should create a module with simple  "feature extraction based" TSC strategies like this, e.g., `feature_extr_based`.

Other shorthands that would be great to have:
- [ ]  `TabularizerClassifier`, using the tabluarizer; a bonus feature would be using its potential "good first issue" extension to be created, the binner #242
- [x] `MatrixProfileClassifier` using the matrix profile
- [x] `SignatureClassifier` - using signature - this already exists in @jambo6's work, so just needs to be moved into the same "TSC type" sub-folder
- [ ] `SevenNumberClassifier` - using the seven-number-summary of the series (quartiles, mean, variance). A nice feature would be the ability to specify which of these to use, or more "sample summaries" like kurtosis or other percentiles; this should not use any sequential features, only features where the order does not matter.

All of these will have parameters: `estimator` - a `sklearn` classifier; and any parameters that come from the feature extractor, without additional nesting level (that you would get from `Pipeline`).

### PR DESCRIPTION
**Pull request by MatthewMiddlehurst**
Implements new feature based pipelines for summary statistics, random intervals (with a new transformer to generate them) and a TSFresh based pipeline from an upcoming workshop paper by me and @TonyBagnall. We can probably close #1063 after this unless there are any objections.

There is quite a lot going on here, but the majority of new lines and file changes are just tests.

#### Does your contribution introduce a new dependency? If yes, which one?

No

#### What should a reviewer concentrate their feedback on?

The new estimators, they are relatively simple and most are basic transformer > classifier pipelines.
Files of interest are:

sktime/classification/feature_based/_fresh_prince.py
sktime/classification/feature_based/_random_interval_classifier.py
sktime/classification/feature_based/_summary_classifier.py
sktime/transformations/panel/random_intervals.py

### CODE REVIEW

**Review by TonyBagnall — APPROVED** [review]
this looks fine, and we can work it into the new classification notebook I'm planning with pure pipeline examples and the static implementations. Needs to wait until after the release of 0.9 though, which should happen tonight (weds 8/12/21)

**Review by fkiraly — CHANGES_REQUESTED** [review]
Question: why are we introducing n classifiers following the pattern [transformer][classifier], but not a generic pipeline object that implements this pattern?

Makes sense to have shorthands, but it would be more robust if the shorthand is a shorthand for the composable pipeline?

**Comment by MatthewMiddlehurst** [comment]
Is the sklearn Pipeline not a suitable generic object that follows this pattern? Unsure if a middle step is necessary. 

```
fresh_prince = FreshPRINCE()
```
Is a shorthand for:

```
fresh_prince = Pipeline([
    ("transform", TSFreshFeatureExtractor(default_fc_parameters="comprehensive")),
    ("classifier", RotationForest(n_estimators=200)),
])
```

We could have a TransformClassifier:

```
fresh_prince = TransformClassifier(
    transform=TSFreshFeatureExtractor(default_fc_parameters="comprehensive"),
    classifier=RotationForest(n_estimators=200),
)
```

But I don't think it would add much value beyond people being unfamiliar with pipelines, possibly. I think in the upcoming example rework for classification, more focus will be put on them.

It should be pretty easy to put together, however. I don't mind doing so for this PR if people think it is a good addition.

**Comment by fkiraly** [comment]
@MatthewMiddlehurst, that's almost what I mean.

Though what I mean is one of the following two

(a) a class factory along the lines of

```python
FreshPRINCE = PipelineFactory([
    ("transform", TSFreshFeatureExtractor(default_fc_parameters="comprehensive")),
    ("classifier", RotationForest(n_estimators=200)),
])
```

or (b) internally you make use on `sklearn` `Pipeline`, instead of re-implementing and copy-pasting the pipeline logic three times, e.g.,

```python
def __init__(stuff):
    more_stuff
    self.pipeline_ = Pipeline([
        ("transform", TSFreshFeatureExtractor(default_fc_parameters="comprehensive")),
        ("classifier", RotationForest(n_estimators=200)),
    ])

def _fit(self, X, y):
    foo
    bar
    self.pipeline_.fit(X, y)
    return self
````
or similar.

**Comment by fkiraly** [comment]
looking at this again, though - I wonder whether it's possible to avoid the boilerplate code in there. 

So I'd downgrade my above comments to more something for consideration, not a blocker.

File: sktime/classification/feature_based/tests/test_random_interval_classifier.py
Comment (fkiraly) [inline sktime/classification/feature_based/tests/test_random_interval_classifier.py:174]: please remove commented out code

File: sktime/classification/feature_based/tests/test_summary_classifier.py
Comment (fkiraly) [inline sktime/classification/feature_based/tests/test_summary_classifier.py:161]: please remove commented out code

**Review by fkiraly — CHANGES_REQUESTED** [review]
my requests would be:
* please remove dead/commented code
* consider the "boilerplate" issue - if you can argue that there are no DRYer alternatives, I'm happy

**Comment by MatthewMiddlehurst** [comment]
Pipelines get a bit restrictive when you want to add additional functionality. Or in some cases just would not work, i.e. for the summary classifier the default multivariate output would not be compatible for any sklearn classifier.

For sure, some of them could be Pipeline wrappers though.

File: sktime/classification/feature_based/tests/test_random_interval_classifier.py
Comment (MatthewMiddlehurst) [reply sktime/classification/feature_based/tests/test_random_interval_classifier.py:174]: That is the code for generating the results we compare against, left in with the reasoning that it would be very annoying to have to re-implement it every time the results have to be redone after a legitimate change.

I'm against removing it completely, but we could consider formalising it a bit more, i.e. relocating the print function to utils and cleaning up the main portion a bit.

**Comment by TonyBagnall** [comment]
> Question: why are we introducing n classifiers following the pattern [transformer][classifier], but not a generic pipeline object that implements this pattern?
> 
> Makes sense to have shorthands, but it would be more robust if the shorthand is a shorthand for the composable pipeline?

we already have a pipeline, or do you mean a wrapper for a pipeline where transformer and classifier are specified in constructor? Something like a PipelineClassifier? Seems a bit superfluous and I thought you liked using pipelines directly, but I see no reason why not have one.

I agree its best to not have commented out code. Maybe put it all in a directory in contrib called regenerate tests or somesuch?

**Comment by MatthewMiddlehurst** [comment]
RE: Commented out sections for tests, that is a bigger issue than this PR (same pattern follows for most tests), but I don't mind doing it in a separate PR with all the others.

**Review by TonyBagnall — DISMISSED** [review]

**Comment by TonyBagnall** [comment]
> RE: Commented out sections for tests, that is a bigger issue than this PR (same pattern follows for most tests), but I don't mind doing it in a separate PR with all the others.

lets discuss at the dev meeting, I'll put together an agenda

**Review by TonyBagnall — APPROVED** [review]

**Review by TonyBagnall — DISMISSED** [review]

**Comment by TonyBagnall** [comment]
@fkiraly could you rereview in the light of comments above please? ta

**Comment by TonyBagnall** [comment]
we agreed at the dev meeting to centralise the commented out code in a separate PR. @fkiraly does this answer your issues?

**Comment by fkiraly** [comment]
> @fkiraly does this answer your issues?

Yes, happy if there's a separate issue that says what needs be doing.

What did you decide on the boilerplate/DRY issue?

**Comment by MatthewMiddlehurst** [comment]
I still don't really see an issue with the current set-up, issues such as SummaryClassifier multivariate output and extra functionality like train estimates in FreshPrince would not lend well to a generic pipeline.

Im not against having a generic pipeline wrapper for users. Using such a wrapper in the current pipelines would require a bit of unwrapping for the above cases, however.

File: sktime/classification/feature_based/_fresh_prince.py
Comment (chrisholder) [inline sktime/classification/feature_based/_fresh_prince.py:21]: I believe you need an _ to reference so it should be:

`the input data using the TSFresh [1]_`

**Review by chrisholder — APPROVED** [review]

**Comment by TonyBagnall** [comment]
so to summarise, we thought a wrapper containing a configurable pipeline was a layer of wrapping too far, and it is in fact preferable to encourage users to use a Pipeline, since sklearn users are familiar with them. There is not that much code repetition and not all classifiers could use the base class.

**Review by fkiraly — APPROVED** [review]
questions answered, looks fine to me now.

Can you please remove commented out code before you merge?

**Comment by TonyBagnall** [comment]
> questions answered, looks fine to me now.
> 
> Can you please remove commented out code before you merge?

this is deferred to here https://github.com/alan-turing-institute/sktime/issues/1765

**Pull request merged**
